### ISSUE TITLE
Ask-and-tell interface?

### ISSUE THREAD

**Issue by AlexanderFabisch** [scikit-optimize/scikit-optimize#68]
Hi, I just discovered this project. I wonder whether it is really the goal to provide only a scipy-like interface or whether you think it would be possible to provide an [ask-and-tell](https://www.lri.fr/~hansen/collette2010Chap14.pdf) interface, too. That would be much more convenient for use cases in which the optimization process is controlled actually by the objective function.

**Comment by glouppe** [comment]
Am I correct in understanding that it is like a warm start interface?

**Comment by AlexanderFabisch** [comment]
If warm start meant that the state of the optimization algorithm is preserved when it is not running that would be true. The scipy-like implementation based on ask and tell would look like

```python
def myoptimizer_minimize(func, bounds, ...):
    optimizer = MyOptimizer(bounds, ...)

    while not optimizer.stop():
        x = optimizer.ask()
        y = func(x)
        optimizer.tell(x, y)
    ...
```

You could return the optimizer and do a warm start with it.

**Comment by AlexanderFabisch** [comment]
The use case that I have in mind is this:

```python
opt = MyOptimizer(...)
sim = MyComplexSimulation(...)
func = MyFunction(...)

while not opt.stop():
    x = optimizer.ask()

    # Here we start a complex simulation or an experiment in the real world.
    # This is usually not only a single function call and I would like to have
    # the control in my hand and not in the hand of the optimizer.
    result = sim.do_simulation(x, ...)
    y = func(result)

    optimizer.tell(x, y)
```

**Comment by MechCoder** [comment]
Interesting, if I understand correctly the idea is to decouple the function call from the optimization strategy?

Would you be able to provide a more specific example? What would be an example of `do_simulation` when you want a GP to fit the `do_simulation + func` call?

**Comment by betatim** [comment]
I like the idea. You could also use this to support cases where the "function" you want to optimise is a physical process that involves a human in a lab doing something, days later coming back with the answer.

One way of supporting it would be to support passing in a set of parameters `x`s and corresponding `func(x)`s that you somehow obtained previously and run only one iteration that returns the next point at which to sample.

**Comment by AlexanderFabisch** [comment]
> Interesting, if I understand correctly the idea is to decouple the function call from the optimization strategy?
> 
> Would you be able to provide a more specific example? What would be an example of do_simulation when you want a GP to fit the (x, func(do_simulation(x)) call?

An example would be policy search (e.g. [gait optimization](http://wiki.icub.org/images/b/bc/VVV13BayesianGaitOptimization.pdf)). In policy search you have a parameterizable movement (/policy/behavior/...) that you want to maximize. That means you obtain the parameters from the optimizer, initialize your behavior representation with the new set of parameters, run an experiment on a robot or in simulation, compute some reward or fitness value and give it back to the optimization algorithm.

**Comment by glouppe** [comment]
I see. Hmm well I think we could certainly factor the inner code so that it works in this way, while still providing a scipy-like `minimize` function. This would certainly be a good thing to do for the codebase, to harmonize all the common parts.

**Comment by glouppe** [comment]
Maybe something like:

```
class Optimizer:
    def __init__(**kwargs)
    def ask()
    def tell(x, y)
    def run():
        loop on ask+tell

def minimize(**kwargs):
    opt = Optimizer(**kwargs)
    opt.run()
    return opt.state_   # an OptimizeResult object
```

**Comment by betatim** [comment]
Can we keep the current function based API (I'd like that) if we introduce a convention that allows you to pass in previous points and only perform one iteration.

If you pass in a set of previously evaluated points (warm start) we skip the initial random sampling and with `maxiter=1` you will get back the next set of parameters to evaluate. You would have to provide a dummy `func` as well. We could make this more convenient by providing some syntactic sugar for this.

Another option would be to extract the code inside the `for n range(maxiter)` loop into a function that becomes part of the API.

**Comment by AlexanderFabisch** [comment]
Maybe it would be a better idea to have at least a state object for each optimization algorithm that can be returned on request by `minimize`. I can think of black-box optimization algorithms that have a more complex state that is computationally complex to recompute, e.g. CMA-ES.

**Comment by MechCoder** [comment]
I may be wrong but what are the disadvantages of doing this as @betatim suggests assuming we do support the `warm_start` interface?


     init_points = rng.randn(1, )
     for i in range(n_iter):
          res = gp_func(lambda x: x, bounds, x_init, y_init, n_start=0, maxiter=1)
          x_best = res.x_iters[-1]
          y_best = expensive_func(x_best)
          x_init = np.concatenate..
          y_init = np.concatenate...

This seems hacky to me though.

Alternatively, we could also just have a function `return_next_best_x` (worded better) which would return just the next best x value, given a series of previous x and y values, without computing the computationally expensive y. It would be much cleaner to use this in a loop as compared to the previous function. We could also use this inside the `gp_minimize` loop.

By a state object, is there anything that you cannot recover from the OptimizeResult object provided. (i.e, the x sampled at every iteration, the GP model fit etc). I think the more important issue it to avoid computing `func(x)` and allow the control to be passed to the user no?  This can be addressed by the 2nd option, I believe.

**Comment by AlexanderFabisch** [comment]
OK, that seems to be a good solution. OptimizeResult actually captures the whole state of the optimization algorithm.

**Comment by glouppe** [comment]
> I may be wrong but what are the disadvantages of doing this as @betatim suggests assuming we do support the warm_start interface?

This would be enough indeed, but it really does look hackish.

I would prefer an explicit API as sketched above, which would structure our codebase into common components. The scipy-like API would still be there, as shortcuts. (Just like we have `StandardScaler` and `scale` in scikit-learn).

**Comment by MechCoder** [comment]
At some point of time, we should implement @glouppe 's idea. But right now getting the input API high and SMAC working is of highest priority.

**Comment by glouppe** [comment]
Scikit-Optimize now provides a ask-and-tell interface! Closed by #234.
